### ISSUE TITLE
[1.17.11] Game Runs in OpenGL 4.6 Compatibility Profile On AMD Systems with Drivers Starting 22.7.1; Causes Major Performance Issues

### ISSUE THREAD

**Issue by kurazarrh** [anegostudios/VintageStory-Issues#2454]
### Game Version

1.17.11

### Platform

Windows

### Modded

Vanilla

### SP/MP

Singleplayer

### Description

For Windows systems running AMD video cards, using driver version 22.7.1 or newer causes the game to run in OpenGL 4.6 Compatibility Profile instead of OpenGL 3.3 Core Profile.

This appears to be causing major performance issues (see attached screenshots).

This may be due to how AMD overhauled their OpenGL support in driver version 22.7.1; it may necessitate an update to how Vintage Story initializes OpenGL.

### How to reproduce

1) Install an AMD video card or use a laptop with an AMD video card.
2) Install AMD drivers version 22.7.1 or newer.
3) Run Vintage Story.
4) Load into a world.
5) Press Ctrl+F3.

### Screenshots

![Vintage_Story_AMD_Driver_22 5 1](https://user-images.githubusercontent.com/29897889/223886890-39cf2a6d-180d-420b-8f90-d9cbc0b22ed2.png)
![Vintage_Story_AMD_Driver_22 11 2](https://user-images.githubusercontent.com/29897889/223886894-dc8922e3-6fa6-41f4-967e-9c50582e2d5d.png)

**Comment by tyronx** [comment]
the game requests a ogl 3.3 context. If the driver ignores that, there's not really anything i can do about that, as far as i know.

**Comment by kurazarrh** [comment]
It may be worth looking into whether there's a new method for requesting the required OGL version that's specific to AMD drivers starting in 22.7.1. It appears that another block game (one that shall not be mentioned here) is able to request and receive the correct OGL version on this and newer driver versions, which is why I suspect the request method has been changed.

**Comment by tyronx** [comment]
i much rather prefer amd not bork their drivers 😏

**Comment by tyronx** [comment]
Generally speaking - graphics card vendors optimize performance on a per-game basis. If they make minecraft run faster, it shouldn't affect other games.

**Comment by kurazarrh** [comment]
Are you suggesting they put in some code in their API that only allows Minecraft to request and receive its desired OpenGL version, while preventing other applications from getting the OpenGL version they need?

**Comment by kurazarrh** [comment]
In case it's something on AMD's end (which it may well be), I've also submitted a ticket on their forums (I don't have access to post in their OpenGL forum). If you want to take a look at it or if you can shed any insight into the issue, you can check it out here: https://community.amd.com/t5/drivers-software/bug-report-opengl-application-runs-in-4-6-compatibility-profile/m-p/592249#M171376

**Comment by kurazarrh** [comment]
@tyronx I got a response from AMD on the issue, and they're looking into it as a possible issue with AMD's drivers. Would you be able/willing to provide a game key to their developers if they were to reach out to you regarding this issue, for testing potential fixes?

**Comment by tyronx** [comment]
sure, i can do that

**Comment by kurazarrh** [comment]
@tyronx Awesome, thank you! Here's a link to the bug again (they moved it to a different forum, so new link); they're asking for how to get the game. https://community.amd.com/t5/opengl-vulkan/bug-report-opengl-application-runs-in-4-6-compatibility-profile/m-p/593441#M4581

If you don't want to have to make an AMD account and would rather have me pass along the game key, I can do that. Let me know how I can help. Thanks!

**Comment by tyronx** [comment]
Oh, I'm not allowed to post there. Could you ask them to give me acccess, here is my account - https://community.amd.com/t5/user/viewprofilepage/user-id/260781

A direct contact from me with the devs is probably going to be more efficient.

**Comment by kurazarrh** [comment]
Absolutely! I figured that would be preferable. They're even asking for the game key now, so it's perfect timing! XD

**Comment by kurazarrh** [comment]
> Oh wow, you absolute gigachad lol. I would not have not expected that we would be able to get in contact with an actual AMD dev. I'll add a post to your thread, I seem to already have an account there

I figured if Vintage Story was the only game having OpenGL issues, it wouldn't get much attention, but I spent some time snooping around their forums and found a lot of people complaining about awful OpenGL performance, but no one had quite hit on the OpenGL applications getting stuck with the 4.6 Compatibility Profile. I still have no explanation for why That Other Block Game gets its requested version of OpenGL (3.2) but not other applications. Maybe the Java platform has some workaround... or more likely, it relies on another bug in order to work correctly... XD

But yeah, I have a very large interest in helping to get this bug resolved--I make a LOT of Vintage Story videos over on my YouTube channel, and it's a ton of fun. I want to keep doing it, so I'm going to rattle whichever cage is going to get results! :D

**Comment by kurazarrh** [comment]
> Oh, I'm not allowed to post there. Could you ask them to give me acccess, here is my account - https://community.amd.com/t5/user/viewprofilepage/user-id/260781
> 
> A direct contact from me with the devs is probably going to be more efficient.

I've posted a request to dipak to whitelist your account in the OpenGL forums. Unfortunately, since we're both below the "Novice II" rank, dipak isn't able to PM us, but I'll check back in later today and see if he's replied to the comment and will let you know here, unless you happen to notice it first.

**Comment by kurazarrh** [comment]
@tyronx Oh holy cow. Dipak replied and has already whitelisted you. You should be able to post on that forum now. Looks like he's also looking in to how to force-enable the PM option for both our accounts.

**Comment by Parovchenko** [comment]
Hello. So is there any solution? I have a processor with integrated Vega 7 graphics. And when I use the latest versions of drivers (currently 23.4.1 is installed), when I go to a server or a single game where there are a large number of buildings, I get a complete crash of the game, a black screen (while the game continues work) or textures disappear.

**Comment by kurazarrh** [comment]
This bug has been sent to AMD's OpenGL team. They are still looking at it, as far as I am aware. Recommendation is to use driver 22.5.1, though beware that you should also use DDU to uninstall drivers instead of letting AMD's driver package do it, unless you want to risk bricking your Windows installation.
________________________________

**Comment by kurazarrh** [comment]
The selection of the wrong OpenGL version happens regardless whether OBS is running. OBS's presence seems to magnify the issue rather dramatically, but even without it running, my frame time graph is a LOT bumpier when running with the new AMD drivers vs. the old ones.

To be fair, OBS and Vintage Story have a weird relationship, anyway. If I set the VS FPS limiter to ~70fps or so WITHOUT OBS running, my framerate gets locked at 45fps. But when I start up OBS (whether before running VS or with the game running), it changes my fps lock to 60fps (probably has to do with OBS telling the GPU it would very much like to have 60fps for recording). It did this on my old Nvidia PC as well, so I don't think that part is related to the display driver. In either case, this weird interaction isn't impacting the ability to record video; the new AMD driver and likely wrong selection of OpenGL version is.
________________________________

**Comment by tyronx** [comment]
Oh okay, thats good to know.

**Comment by JannesEng** [comment]
Also experiencing this issue on AMD system.
Windows 10 22H2 (build 19045.3570)
GPU: AMD 5700XT
CPU: Ryzen 7 3700 X
RAM: 32 Gig

Have to run AMD 22.6.x drivers to have normal performance.

I commented on the initial AMD forum post [kurazarrh](https://github.com/kurazarrh) made, but I do not have access to the developers forum there.

**Comment by Master449** [comment]
Able to reproduce, 7800XT on Ubuntu 24.04 (and Windows 11 verison unsure since its been wiped)

I also can't revert to 22.5.1 since 23.9.1 adds support for the 7000 series :(

**Comment by Master449** [comment]
Okay, I believe this issue can be closed, but I am unsure if you would want to do more testing on more cards.

@Quant-ux @JannesEng Does this issue still affect your systems?

I got in touch with u/AMD_Vik over reddit I asked for an update on the bug report that kurazarrh posted about earlier.

I got back:

> One of our OpenGL driver engineers quickly tested this on v1.19.8. with NV23. Strangely enough, they're seeing a GL 3.3 core context; ~400 FPS at default settings and ~50 FPS at highest.

I personally do not know what changed with my setup, but its now running on 3.3.0 Core.

![image](https://github.com/user-attachments/assets/520031eb-8db1-4314-8f10-9b40ed21bf27)
